core/src/main/java/com/scalar/db/transaction/consensuscommit/CommitHandler.java

```diff
@@ -104,28 +104,30 @@ public void commit(Snapshot snapshot) throws CommitException, UnknownTransaction
     try {
       prepare(snapshot);
     } catch (PreparationException e) {
+      onFailureBeforeCommit(snapshot);
```

Comment:
Actually, the following `abortState()` internally calls a method equivalent to `onFailureBeforeCommit()` in case of `CommitHandlerWithGroupCommit`. So, this change doesn't change the behavior at all. But, I think the callback should be called here for consistent error handling.